simulation/array_simulator.pro

```diff
@@ -51,7 +51,27 @@ PRO array_simulator,vis_arr,vis_weights,obs,status_str,psf,params,jones,error=er
   IF ~silent THEN print,'Calculating beam model'
   psf=beam_setup(obs,status_str,file_path_fhd=file_path_fhd,restore_last=0,silent=silent,timing=t_beam,no_save=0,_Extra=extra)
   IF Keyword_Set(t_beam) THEN IF ~silent THEN print,'Beam modeling time: ',t_beam
-  
+ 
+  ;; Get a selection radius based on the primary beam width  TODO --- Only do this if include_catalog_sources == 0, in which case get it from generate_source_cal_list
+  if not keyword_set(beam_threshold) then beam_threshold = 0.05
+  if keyword_set(allow_sidelobe_sources) THEN beam_threshold = 0.01
+
+  dimension=obs.dimension
+  elements=obs.elements
+  degpix=obs.degpix
+  print, 'Getting beam width'
+  IF N_Elements(beam_arr) LT (n_pol<2) THEN BEGIN 
+      beam_arr=Ptrarr(n_pol<2)
+      FOR pol_i=0,(n_pol<2)-1 DO beam_arr[pol_i]=Ptr_new(beam_image(psf,obs,pol_i=pol_i,square=0)>0.)
+  ENDIF
+  beam=fltarr(dimension,elements)
+  FOR pol_i=0,(n_pol<2)-1 DO beam+=*beam_arr[pol_i]^2.
+  beam=Sqrt(beam/(n_pol<2))
+  beam_primary_i=region_grow(beam,dimension/2.+dimension*elements/2.,threshold=[Max(beam)/2.<beam_threshold,Max(beam)>1.])
+  boundary=find_boundary(beam_primary_i,xsize=dimension, ysize=elements,perim_area=perim_area)
+  select_radius = (degpix*sqrt(perim_area)/2 + 3.)   ;Primary beam radius + 3 degrees
```

Comment:
This new code needs to all be in a separate function. Then you can safely set the defaults for beam_threshold you use above, though you should think about what you are doing if `allow_sidelobe_sources` is set. This calculation will expand the beam, but the radius it produces will *not* actually include the sidelobes. So, either treat the sidelobes properly or (most likely) simply ignore the keyword `allow_sidelobe_sources` for this calculation.

Reply:
Most of this is just a duplicate of what's done in generate_source_cal_list for source selection (which sets defaults for beam_threshold according to allow_sidelobe_sources). The only change is the estimate of the beam radius from that. Where should I put this to act as a separate function?

Reply:
Just write a new function, called something like `select_beam_pixels` and put all of this code in it. The new function should go in `fhd_utils`. Please also see my recent commit that changed `generate_source_cal_list`. You will want to apply those changes to this code as well.

Reply:
Do you want the new function to do just the primary_beam width calculation, or do everything that's done in generate_source_cal_list and keep the width estimation in array_simulator?